Description: Add support for localized distribution searchplugins
Author: Alexander Sack <asac@ubuntu.com>
Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=515232
Forwarded: https://bugzilla.mozilla.org/attachment.cgi?id=399338

Index: mozilla/browser/components/dirprovider/DirectoryProvider.cpp
===================================================================
--- mozilla.orig/browser/components/dirprovider/DirectoryProvider.cpp	2012-03-07 19:47:01.707413688 +0000
+++ mozilla/browser/components/dirprovider/DirectoryProvider.cpp	2012-03-07 19:56:03.147404067 +0000
@@ -35,6 +35,7 @@
  *
  * ***** END LICENSE BLOCK ***** */
 
+#include "nsIChromeRegistry.h"
 #include "nsIDirectoryService.h"
 #include "DirectoryProvider.h"
 
@@ -189,18 +190,17 @@
         array.AppendObject(commonPlugins);
   }
 
-  nsCOMPtr<nsIPrefBranch> prefs(do_GetService(NS_PREFSERVICE_CONTRACTID));
-  if (prefs) {
-
-    nsCOMPtr<nsIFile> localePlugins;
-    rv = searchPlugins->Clone(getter_AddRefs(localePlugins));
-    if (NS_FAILED(rv))
-      return;
+  nsCOMPtr<nsIFile> localePlugins;
+  rv = searchPlugins->Clone(getter_AddRefs(localePlugins));
+  if (NS_FAILED(rv))
+    return;
 
-    localePlugins->AppendNative(NS_LITERAL_CSTRING("locale"));
+  localePlugins->AppendNative(NS_LITERAL_CSTRING("locale"));
 
-    nsCString locale;
-    rv = prefs->GetCharPref("general.useragent.locale", getter_Copies(locale));
+  nsCOMPtr<nsIXULChromeRegistry> chromeReg(do_GetService(NS_CHROMEREGISTRY_CONTRACTID));
+  if (chromeReg) {
+    nsCAutoString locale;
+    chromeReg->GetSelectedLocale(NS_LITERAL_CSTRING("global"), locale);
     if (NS_SUCCEEDED(rv)) {
 
       nsCOMPtr<nsIFile> curLocalePlugins;
@@ -213,9 +213,12 @@
           array.AppendObject(curLocalePlugins);
           return; // all done
         }
-      }
+      } 
     }
+  }
 
+  nsCOMPtr<nsIPrefBranch> prefs(do_GetService(NS_PREFSERVICE_CONTRACTID));
+  if (prefs) {
     // we didn't append the locale dir - try the default one
     nsCString defLocale;
     rv = prefs->GetCharPref("distribution.searchplugins.defaultLocale",
